src/Microsoft.AspNet.Routing/InlineRouteParameterParser.cs

```diff
@@ -12,82 +9,230 @@ namespace Microsoft.AspNet.Routing
 {
     public static class InlineRouteParameterParser
     {
-        // One or more characters, matches "id"
-        private const string ParameterNamePattern = @"(?<parameterName>.+?)";
+        public static TemplatePart ParseRouteParameter([NotNull] string routeParameter)
+        {
+            if (routeParameter.Length == 0)
+            {
+                return TemplatePart.CreateParameter(
+                    name: string.Empty,
+                    isCatchAll: false,
+                    isOptional: false,
+                    defaultValue: null,
+                    inlineConstraints: null);
+            }
 
-        // Zero or more inline constraints that start with a colon followed by zero or more characters
-        // Optionally the constraint can have arguments within parentheses
-        //      - necessary to capture characters like ":" and "}"
-        // Matches ":int", ":length(2)", ":regex(\})", ":regex(:)" zero or more times
-        private const string ConstraintPattern = @"(:(?<constraint>.*?(\(.*?\))?))*";
+            var startIndex = 0;
+            var endIndex = routeParameter.Length - 1;
 
-        // A default value with an equal sign followed by zero or more characters
-        // Matches "=", "=abc"
-        private const string DefaultValueParameter = @"(?<defaultValue>(=.*?))?";
+            var isCatchAll = false;
+            var isOptional = false;
 
-        private static readonly Regex _parameterRegex = new Regex(
-           "^" + ParameterNamePattern + ConstraintPattern + DefaultValueParameter + "$",
-            RegexOptions.CultureInvariant | RegexOptions.IgnoreCase);
+            if (routeParameter[0] == '*')
+            {
+                isCatchAll = true;
+                startIndex++;
+            }
 
-        public static TemplatePart ParseRouteParameter([NotNull] string routeParameter)
-        {
-            var isCatchAll = routeParameter.StartsWith("*", StringComparison.Ordinal);
-            var isOptional = routeParameter.EndsWith("?", StringComparison.Ordinal);
+            if (routeParameter[endIndex] == '?')
+            {
+                isOptional = true;
+                endIndex--;
+            }
+
+            var currentIndex = startIndex;
 
-            routeParameter = isCatchAll ? routeParameter.Substring(1) : routeParameter;
-            routeParameter = isOptional ? routeParameter.Substring(0, routeParameter.Length - 1) : routeParameter;
+            // Parse parameter name
+            var parameterName = string.Empty;
 
-            var parameterMatch = _parameterRegex.Match(routeParameter);
-            if (!parameterMatch.Success)
+            while (currentIndex <= endIndex)
```

Comment:
I could theoretically roll this into this into a single state machine along with the constraint parsing - but this made it slightly easier to refactor the lot into a separate method.